src/hal/src/display/mod.rs

```diff
@@ -0,0 +1,190 @@
+    PerPixelPremultiplied
+}
+
+impl Default for DisplayPlaneAlpha
+{
+    fn default() -> Self { Self::Opaque }
+}
+
+// This implementation is done to ignore differences on the value in DisplayPlaneAlpha::Global
+impl PartialEq for DisplayPlaneAlpha {
+    fn eq(&self, other: &Self) -> bool {
+        match (self,other)
+        {
+            (DisplayPlaneAlpha::Opaque,DisplayPlaneAlpha::Opaque)=>true,
+            (DisplayPlaneAlpha::Global(_),DisplayPlaneAlpha::Global(_))=>true,
+            (DisplayPlaneAlpha::PerPixel,DisplayPlaneAlpha::PerPixel)=>true,
+            (DisplayPlaneAlpha::PerPixelPremultiplied,DisplayPlaneAlpha::PerPixelPremultiplied)=>true,
+            _=>false
+        }
+    }
+}
+
+/// Error occurring while creating a display plane.
+#[derive(Clone, Debug, PartialEq, thiserror::Error)]
+pub enum DisplayModeError {
+    /// Display error.
+    #[error(transparent)]
+    OutOfMemory(#[from] crate::device::OutOfMemory),
+    /// Unsupported resolution and refresh rate combination
+    #[error("Unsupported resolution and refresh rate combination")]
+    UnsupportedDisplayMode,
+}
+
+/// Error occurring while creating a display plane surface.
+#[derive(Clone, Debug, PartialEq, thiserror::Error)]
+pub enum DisplayPlaneSurfaceError {
+    /// Display error.
+    #[error(transparent)]
+    OutOfMemory(#[from] crate::device::OutOfMemory),
+    /// Unsupported resolution and refresh rate combination
+    #[error("Unsupported parameters used")]
+    UnsupportedParameters,
+}
+
+/**
+Representation of a display
+*/
+#[derive(Debug)]
+pub struct Display<'a,B: Backend>
+{
+    /// The adapter.
+    pub adapter: &'a adapter::Adapter<B>,
+    /// The display handle.
+    pub handle: B::Display,
+    /// General information about this display.
+    pub info: DisplayInfo
+}
+
+/**
+General information about the a [DisplayMode][DisplayMode].
+*/
+#[derive(Debug)]
+pub struct DisplayMode<'a,B: Backend>
+{
+    /// The display
+    pub display: &'a Display<'a,B>,
+    /// The display mode handle
+    pub handle: B::DisplayMode,
+    /// Resolution
+    pub resolution: (u32,u32),
+    /// Refresh rate
+    pub refresh_rate: u32
+}
+
+/**
+Representation of a plane
+*/
+#[derive(Debug)]
+pub struct Plane<'a,B: Backend>
+{
+    /// The adapter.
+    pub adapter: &'a adapter::Adapter<B>,
+    /// The plane handle.
+    pub handle: B::Plane,
+    /// The current index on the z stack.
+    pub z_index: u32
+}
+
+
+/**
+Represent a combination of [display mode][DisplayMode] (so [display][Display] and resolution) and a plane
+*/
+#[derive(Debug)]
+pub struct DisplayPlane<'a,B: Backend>
+{
+    /// Display mode
+    pub display_mode: &'a DisplayMode<'a,B>,
+    /// Plane index
+    pub plane: &'a Plane<'a,B>,
+    /// Supported alpha capabilities
+    pub supported_alpha: Vec<DisplayPlaneAlpha>,
+    /// The minimum source rectangle offset supported by this plane using the specified mode.
+    pub min_src_position: (i32,i32),
```

Comment:
looks like we need new types for the `Extent` and `Origin` here, or something like this

Reply:
Sure, but i have noticed that there is `hal::image::Extent` that is a combination of width, height and depth so there could be some name conflict. Maybe i could just follow the Vulkan name scheme like Extent2D,Extent3D,Offset2D and Offset3D and put everything in something like `hal::common` mod. What do you think?

Reply:
I'm not worried about the name conflict. We aren't re-exporting everything from the root, so it's up to users to decide how to resolve this. All the user code I write, at least, would not be affected.

> Maybe i could just follow the Vulkan name scheme like Extent2D,Extent3D,Offset2D and Offset3D and put everything in something like hal::common mod. What do you think?

Currently we have `hal::window::Extent2D`... How do you feel about putting your code into the `window` module instead of having a separate one? It has the stuff about windowing, surface, and presentation. If you feel that a better isolation of code would be needed, we can have submodules, like `src/window/display.rs`, that are potentially re-exported from the `window`.

Reply:
> How do you feel about putting your code into the `window` module instead of having a separate one?

I didn't noticed there was `hal::window::Extent2D`, so i can simply add a `Offset2D` structure below.